library/common/http/dispatcher.cc

```diff
@@ -33,6 +34,13 @@ void Dispatcher::DirectStreamCallbacks::encodeHeaders(const ResponseHeaderMap& h
   ENVOY_LOG(debug, "[S{}] response headers for stream (end_stream={}):\n{}",
             direct_stream_.stream_handle_, end_stream, headers);
 
+  uint64_t response_status = Utility::getResponseStatus(headers);
+
+  // Record received status to report success or failure stat in
+  // Dispatcher::DirectStreamCallbacks::closeRemote. Not reported here to avoid more complexity with
+  // checking end stream, and the potential for stream reset, thus resulting in mis-reporting.
+  observed_success_ = CodeUtility::is2xx(response_status);
```

Comment:
Why not just emit the stat here?

Reply:
Looks like this is explained via the docs directly above?

Reply:
@junr03 Can correct me if I'm wrong (maybe that'll be best anyway!) but the intent here is to get a high level stat on when the application side is able to receive a success or failure. 

The reason we don't track it here is because we could run into other stream events where the application side wouldn't necessarily get a "complete request" but we'll be tracking a success here.

Reply:
Yep, the intention to not increase here is that we could still receive a reset or a cancel if the on headers call is not final. So instead of increasing in this function in the dispatch block for the terminal case, and in on complete for the other cases I decided to just increase when on complete is dispatched.

Reply:
I guess what I was getting at is you could emit stream_opened or stream_failed here. If canceled happens later that can still be reported and you (arguably?) have more insight, rather than less, with (maybe?) less complexity since you don't need state. But I guess it all depends on what you actually want captured.

I missed the comment. :)

Reply:
I probably should have been more clear also - it's not that big a deal, and I am fine with the current approach. was mostly just wondering if the extra state was really necessary.